File: benchmarks/semantic_search/fixtures/go-workflow-discovery-v1/internal/discovery/filters.go
package discovery

import "context"

type Signal struct {
	Labels map[string]string
}

type DiscoveryFilters struct {
	Component  string
	Environment string
	Severity   string
}

func FiltersFromSignal(signal Signal) DiscoveryFilters {
	return DiscoveryFilters{
		Component:   signal.Labels["component"],
		Environment: signal.Labels["environment"],
		Severity:    signal.Labels["severity"],
	}
}

func ApplyLabelFilters(ctx context.Context, workflows []Workflow, filters DiscoveryFilters) []Workflow {
	_ = ctx
	filtered := make([]Workflow, 0, len(workflows))
	for _, workflow := range workflows {
		if filters.Component != "" && workflow.Labels["component"] != filters.Component {
			continue
		}
		if filters.Environment != "" && workflow.Labels["environment"] != filters.Environment {
			continue
		}
		if filters.Severity != "" && workflow.Labels["severity"] != filters.Severity {
			continue
		}
		filtered = append(filtered, workflow)
	}
	return filtered
}

File: benchmarks/semantic_search/fixtures/python-workflow-discovery-v1/discovery/filters.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass
class Signal:
    labels: dict[str, str]


@dataclass
class DiscoveryFilters:
    component: str = ""
    environment: str = ""
    severity: str = ""


def filters_from_signal(signal: Signal) -> DiscoveryFilters:
    return DiscoveryFilters(
        component=signal.labels.get("component", ""),
        environment=signal.labels.get("environment", ""),
        severity=signal.labels.get("severity", ""),
    )


def apply_label_filters(workflows: list[object], filters: DiscoveryFilters) -> list[object]:
    expected = {
        key: value
        for key, value in {
            "component": filters.component,
            "environment": filters.environment,
            "severity": filters.severity,
        }.items()
        if value
    }
    return [workflow for workflow in workflows if all(getattr(workflow, "labels", {}).get(k) == v for k, v in expected.items())]
